src/bin/pg_autoctl/pgsql.c

```diff
@@ -1491,12 +1491,21 @@ pgsql_drop_replication_slot(PGSQL *pgsql, const char *slotName)
  * We return how many parameters we filled in paramTypes and paramValues from
  * the nodeArray.
  */
+typedef struct nodesArraysValuesParams
+{
+	int count;
+	Oid types[NODE_ARRAY_MAX_COUNT * 2];
+	char *values[NODE_ARRAY_MAX_COUNT * 2];
+	char data[NODE_ARRAY_MAX_COUNT * 2][PG_LSN_MAXLENGTH];
```

Comment:
maybe split this array in two fields one for LSNs and one for nodeids. I think that makes the code a bit easier to understand.

Reply:
The job of this routine is to prepare arguments for `pgsql_execute_with_params`, using the libpq interface there. The thing is, this routine wants `(const char **)` and we can't use a `char [][]` to the same effect, because of pointer arithmetic and stuff in the C language. That's why we have a `char data[][]` for storage and a `char *values[]` for pointers...

Reply:
I meant something like this:

```suggestion
	char nodeIds[NODE_ARRAY_MAX_COUNT][INT_STRING_MAX_LENGTH];
	char lsns[NODE_ARRAY_MAX_COUNT][PG_LSN_MAXLENGTH];
```

Reply:
Yeah well that could work now that we store the pointers to the data separately. Let me have a look then, I agree it would simplify quite some things.

Reply:
Done now, thanks for the idea.